### PR TITLE
UEFI cross: don't overthink it

### PR DESCRIPTION
**Pull request by alyssais**
###### Description of changes

#226145 was a good spike at getting UEFI cross compilation to work, but we got bogged down in questions about triples and wrappers.  I've come to the conclusion that actually, we don't need a new triple, as the only place where we want to diverge from normal MSVC Windows is where we need to accomodate not having a libc, and that's not something that triples encode (at least not the GNU-style ones we generally use in Nixpkgs).  (It's a bit surprising, but it's true! For example, Linux's build system just uses either -gnu or -musl for each architecture it supports.)  After that realisation, I started implementing UEFI from scratch, and found it became much simpler.  Here, I've built up to UEFI cross support by just making a bunch of small, common sense changes that improve handling of either Windows, or no-libc.  This is all we need to be able to build the same lanzaboote example package from #226145.

The main goal here is to build Rust programs, but having a Rust compiler means building compiler-rt for the cross system, which demonstrates that with these changes Nixpkgs is also capable of building C/C++ for UEFI.

This is part 1 in a series.  The next part will be configuring Rust in Nixpkgs to build UEFI applications in a way that's compatible with our compiler wrappers.  That work is well under way, but it requires discussing some patches with Rust upstream before it makes sense to integrate it into Nixpkgs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/cc-wrapper/default.nix [hunk 1]
```diff
@@ -520,7 +520,7 @@ stdenv.mkDerivation {
     ''
 
     # TODO: categorize these and figure out a better place for them
-    + optionalString hostPlatform.isCygwin ''
+    + optionalString (hostPlatform.isCygwin || targetPlatform.isMSVC) ''
```
Comment (rrbutani) [inline pkgs/build-support/cc-wrapper/default.nix:523]: ah good catch! It did not occur to me that this isn't a UEFI specific limitation but a broader Windows limitation.

Not something we need to resolve for this PR to go through but: [should this conditional include `MinGW`](https://bugs.webkit.org/show_bug.cgi?id=163949)?

**Review by rrbutani — APPROVED** [review]
Thank you for taking this on; this looks much better than our hacks in #226145! :slightly_smiling_face: 

---

> The next part will be configuring Rust in Nixpkgs to build UEFI applications in a way that's compatible with our compiler wrappers. That work is well under way, but it requires discussing some patches with Rust upstream before it makes sense to integrate it into Nixpkgs.

Is the plan to get `rustc` to (when using `lld-link`) emit UNIX style command line args for Windows targets so we don't have to rework `cc-wrapper`/`ld-wrapper`?

**Review by rrbutani — CHANGES_REQUESTED** [review]

**Comment by alyssais** [comment]
> Is the plan to get `rustc` to (when using `lld-link`) emit UNIX style command line args for Windows targets so we don't have to rework `cc-wrapper`/`ld-wrapper`?

The plan is to get rustc to not use `lld-link`, but rather `ld.lld`, like on every other target we support.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 1 above)
Comment (alyssais) [reply pkgs/build-support/cc-wrapper/default.nix:523]: I don't think so, because I can cross-build packages for mingw just fine at the moment.

File: pkgs/development/compilers/llvm/11/compiler-rt/default.nix [hunk 2]
```diff
@@ -74,7 +72,10 @@ stdenv.mkDerivation {
   ];
 
   preConfigure = lib.optionalString stdenv.hostPlatform.isDarwin ''
-    cmakeFlagsArray+=("-DCMAKE_LIPO=$(command -v ${stdenv.cc.targetPrefix}lipo)")
+    cmakeFlagsArray+=(
+      "-DCMAKE_LIPO=$(command -v ${stdenv.cc.targetPrefix}lipo)"
+      "-DCMAKE_C_FLAGS=-nodefaultlibs -ffreestanding"
+    )
```
[the remark this answers is not in the document]
Comment (alyssais) [reply pkgs/development/compilers/llvm/11/compiler-rt/default.nix:78]: Oh whoops yeah, didn't realise this was in a Darwin conditional.

[the remark this answers is not in the document]
Comment (alyssais) [reply pkgs/development/compilers/llvm/11/compiler-rt/default.nix]: > I assume `-nodefaultlibs` was supposed to be gated on `useLLVM` like in the other package sets?

Probably? I'm going to leave figuring that out to whoever ends up working on LLVM unification — I've had quite enough of it for the time being.  For now, I've just added `-ffreestanding` to the existing condition checked for by LLVM 11.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 1 above)
Comment (alyssais) [reply pkgs/build-support/cc-wrapper/default.nix:523]: Oh, I see the linked bug report mentions it just doesn't do anything on MinGW.  So in that case I guess it should just be `targetPlatform.isWindows`.

**Review by rrbutani — APPROVED** [review]

**Comment by amjoseph-nixpkgs** [comment]
> libc, and that's not something that triples encode (at least not the GNU-style ones

- https://git.savannah.gnu.org/cgit/config.git/tree/config.sub#n1718
- https://git.savannah.gnu.org/cgit/config.git/tree/config.guess#n1161

File: lib/systems/examples.nix [hunk 3]
```diff
@@ -221,6 +221,32 @@ rec {
     libc = "newlib";
   };
 
+  #
+  # UEFI
+  #
+
+  # See: https://github.com/rust-lang/rust/pull/56769/files#r241916113
+  i686-uefi = {
+    system = "x86_64-windows";
```
Comment (amjoseph-nixpkgs) [inline lib/systems/examples.nix:230]: This causes `platform.config` to be `i686-unknown-uefi` which is rejected by `config.sub`:

```
$ /nix/store/ayj17jhbiqc1k043ld3nnb3xqs0p22sl-gnu-config-2023-01-21/config.sub i686-unknown-uefi
Invalid configuration `i686-unknown-uefi': OS `uefi' not recognized
```

so please:

```suggestion
    config = "x86_64-w64-mingw32";
```

This is the `gnu-config` triple for this platform.   Please use it.  See https://github.com/NixOS/nixpkgs/pull/226145#pullrequestreview-1403683064

File: lib/systems/examples.nix [hunk 4]
```diff
@@ -221,6 +221,32 @@ rec {
...
+    rust.config = "i686-unknown-uefi";
+    useLLVM = true;
+    libc = null;
+  };
+
+  x86_64-uefi = {
+    system = "x86_64-windows";
```
Comment (amjoseph-nixpkgs) [inline lib/systems/examples.nix:237]:
```suggestion
    config = "x86_64-w64-mingw32";
```

File: lib/systems/examples.nix [hunk 5]
```diff
@@ -221,6 +221,32 @@ rec {
...
+    rust.config = "x86_64-unknown-uefi";
+    useLLVM = true;
+    libc = null;
+  };
+
+  aarch64-uefi = {
+    system = "aarch64-windows";
```
Comment (amjoseph-nixpkgs) [inline lib/systems/examples.nix:244]:
```suggestion
    config = "aarch64-w64-mingw32";
```

File: lib/systems/default.nix [hunk 6]
```diff
@@ -50,6 +50,7 @@ rec {
         else if final.isFreeBSD             then "fblibc"
         else if final.isNetBSD              then "nblibc"
         else if final.isAvr                 then "avrlibc"
+        else if final.isGhcjs               then null
```
[the remark this answers is not in the document]
Comment (alyssais) [reply lib/systems/default.nix:53]: Did you look at it in the context of its commit, 9fa63d3756cc83e14c6ceac9d80e4bfd036a593f?  The commit removes special handling of GHCJS in `libcCrossChooser` so it can be used by other `libc = null` targets as well, and this is part of that refactoring.

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -20400,7 +20401,6 @@ with pkgs;
     else if name == "nblibc" then targetPackages.netbsdCross.libc or netbsdCross.libc
     else if name == "wasilibc" then targetPackages.wasilibc or wasilibc
     else if name == "relibc" then targetPackages.relibc or relibc
-    else if stdenv.targetPlatform.isGhcjs then null
```
Comment (amjoseph-nixpkgs) [inline pkgs/top-level/all-packages.nix:20403]: Was this deletion accidental?  Seems unrelated.

**Comment by alyssais** [comment]
> > libc, and that's not something that triples encode (at least not the GNU-style ones
> 
>     * https://git.savannah.gnu.org/cgit/config.git/tree/config.sub#n1718
> 
>     * https://git.savannah.gnu.org/cgit/config.git/tree/config.guess#n1161

If you look at a little bit more of my sentence:

> not having a libc, and that's not something that triples encode

you'll see that I was not saying "triples don't encode which libc is in use".  I am making a more specific claim, which is that triples do not encode _absence_ of libc.  I looked at the same files you linked to verify I was correct about that.  I think it's strange too, but that's how it is.

**Review by ghost — COMMENTED** [review]
LGTM except for the triple choice.

File: lib/systems/examples.nix (hunk 3 above)
Comment (alyssais) [reply lib/systems/examples.nix:230]: > causes `platform.config` to be `i686-unknown-uefi` which is rejected by `config.sub`:
> 
> ```
> $ /nix/store/ayj17jhbiqc1k043ld3nnb3xqs0p22sl-gnu-config-2023-01-21/config.sub i686-unknown-uefi
> Invalid configuration `i686-unknown-uefi': OS `uefi' not recognized
> ```

No it doesn't:

```
nix-repl> (lib.systems.elaborate "i686-windows").config
"i686-pc-windows-msvc"
```

config.sub won't recognise this either, but that's because GNU doesn't support the MSVC ABI, so there's no correct triple to use here.  This is correct for LLVM, which is what we actually use with these system definitions.

Are you just copying these comments from the other PR without checking whether they apply to this one?  It seems like you're arguing against things that this PR isn't trying to do…

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (alyssais) [reply pkgs/top-level/all-packages.nix:20403]: No, as I [already explained](https://github.com/NixOS/nixpkgs/pull/228374#discussion_r1178957343), I've removed special handling of GHCJS here, so that other platforms can also have libc = null.

**Comment by amjoseph-nixpkgs** [comment]
> which is that triples do not encode _absence_ of libc.

That is a much clearer way of putting it, but I think it's moot: for binaries with no libc that aren't kernels, this field still encodes the ABI.

> the only place where we want to diverge from normal MSVC Windows is where we need to accomodate not having a libc

The distinctive part of UEFI is its calling convention, i.e. its ABI.

Are you sure that UEFI's calling convention is exactly identical to Windows'?  On [ARM it is definitely totally different](https://wiki.osdev.org/UEFI#Calling_Conventions).

**Comment by alyssais** [comment]
> Are you sure that UEFI's calling convention is exactly identical to Windows'? On [ARM it is definitely totally different](https://wiki.osdev.org/UEFI#Calling_Conventions).

I know that [Rust just tells LLVM it's compiling for Windows](https://github.com/rust-lang/rust/pull/56769/files#r241916113), so I was working based on that.  Do you think they're wrong to do that?

**Comment by alyssais** [comment]
Okay, you know what? The example is not necessary, because people can specify what system they want to use when they import Nixpkgs, and I really just wanted to do a few simple package fixes to get the ball rolling.  So I'm going to drop the examples from this PR, and then hopefully we can agree that the package fixes make sense, and if somebody else wants to take up adding examples, they can do that without me.

**Comment by amjoseph-nixpkgs** [comment]
> I know that [Rust just tells LLVM it's compiling for Windows](https://github.com/rust-lang/rust/pull/56769/files#r241916113), so I was working based on that.

That link is talking specifically about `x86_64`.

**Comment by alyssais** [comment]
Examples dropped.

**Review by rrbutani — APPROVED** [review]

**Review by Ericson2314 — APPROVED** [review]
Looked through commit by commit, and everything looks good!

**Review by ghost — APPROVED** [review]

**Review by RaitoBezarius — APPROVED** [review]
Thank you so much :)
